src/System.Private.ServiceModel/tests/Common/Infrastructure/xunit/WcfTestCase.cs

```diff
@@ -44,13 +50,48 @@ internal WcfTestCase(IXunitTestCase testCase, string skippedReason = null, bool
 
         public void Deserialize(IXunitSerializationInfo info) { _testCase.Deserialize(info); }
 
-        public Task<RunSummary> RunAsync(
+        public async Task<RunSummary> RunAsync(
             IMessageSink diagnosticMessageSink, IMessageBus messageBus, object[] constructorArguments,
             ExceptionAggregator aggregator, CancellationTokenSource cancellationTokenSource)
         {
-            return _isTheory
-                ? new XunitTheoryTestCaseRunner(this, DisplayName, _skippedReason, constructorArguments, _diagnosticMessageSink, messageBus, aggregator, cancellationTokenSource).RunAsync()
-                : new XunitTestCaseRunner(this, DisplayName, _skippedReason, constructorArguments, TestMethodArguments, messageBus, aggregator, cancellationTokenSource).RunAsync();
+            ConcurrentQueue<EventWrittenEventArgs> events = new ConcurrentQueue<EventWrittenEventArgs>();
+	          s_testListener._eventWritten = events.Enqueue;
+
+            RunSummary runsummary = await (_isTheory ? new XunitTheoryTestCaseRunner(this, DisplayName, _skippedReason, constructorArguments, _diagnosticMessageSink, messageBus, aggregator, cancellationTokenSource).RunAsync()
+                : new XunitTestCaseRunner(this, DisplayName, _skippedReason, constructorArguments, TestMethodArguments, messageBus, aggregator, cancellationTokenSource).RunAsync());
+
+            if (runsummary.Failed > 0)
+            {
+                s_testListener._eventWritten = null;
```

Comment:
>s_testListener._eventWritten = null [](start = 16, length = 35)

Unless I misunderstand the code, I think this code to set _eventWritten to null should be outside this if block.  We don't want to leave a static member initialized to a non-null Action after the test is complete, regardless if we pass or fail.  As mentioned elsewhere, I think _eventWritten should be a property, not a field.